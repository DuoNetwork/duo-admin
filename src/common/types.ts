import { ChildProcess } from 'child_process';
export * from '../../../duo-contract-wrapper/src/types';

export interface IBaseMarketData {
	source: string;
	base: string;
	quote: string;
	timestamp: number;
}

export interface ITrade extends IBaseMarketData {
	id: string;
	price: number;
	amount: number;
}

export interface IPriceFix extends IBaseMarketData {
	price: number;
	volume: number;
}

export interface IPrice extends IBaseMarketData {
	period: number;
	open: number;
	high: number;
	low: number;
	close: number;
	volume: number;
}

export interface IBaseEvent {
	contractAddress: string;
	timestamp: number;
	transactionHash: string;
	blockNumber: number;
}

export interface IAcceptedPrice extends IBaseEvent {
	price: number;
	navA: number;
	navB: number;
}

export interface ITotalSupply extends IBaseEvent {
	tokenA: number;
	tokenB: number;
}

export interface IConversion extends IBaseEvent {
	type: string;
	eth: number;
	tokenA: number;
	tokenB: number;
	fee: number;
	pending?: boolean;
	reverted?: boolean;
}

export interface IStatus {
	process: string;
	timestamp: number;
}

export interface IPriceStatus extends IStatus {
	price: number;
	volume: number;
}

export interface INodeStatus extends IStatus {
	block: number;
}

export interface IAccount {
	address: string;
	privateKey: string;
}

export interface IKey {
	publicKey: string;
	privateKey: string;
}

export interface ISqlAuth {
	host: string;
	user: string;
	password: string;
}

export interface IOption {
	forceREST: boolean;
	pair: string;
	sources: string[];
	exSources: string[];
	assets: string[];
	live: boolean;
	dbLive: boolean;
	server: boolean;
	dynamo: boolean;
	aws: boolean;
	gcp: boolean;
	azure: boolean;
	force: boolean;
	gasPrice: number;
	gasLimit: number;
	source: string;
	event: string;
	provider: string;
	period: number;
	base: string;
	quote: string;
	contractType: string;
	tenor: string;
}

export interface ISourceSettings {
	priceInversed: boolean;
	quoteInversed: boolean;
	separator: string;
	isLowercase: boolean;
	tradesInterval: number;
	supportWS: boolean;
	filterByTimestamp: boolean;
	wsLink: string;
}

export interface ISourceAsset {
	base: {
		[code: string]: boolean;
	};
	quote: {
		[code: string]: boolean;
	};
	mapping: string;
	name: string;
}
export interface ISource {
	assets: {
		[code: string]: ISourceAsset;
	};
	settings: ISourceSettings;
}

export interface IBitfinexRawTradeWS {
	timestamp: number;
	seq: string;
	id: number;
	price: string;
	amount: string;
}

export interface IBitstampRawTradeRest {
	date: number;
	tid: number;
	price: number;
	type: string;
	amount: number;
}

export interface IBitstampRawTradeWs {
	microtimestamp: string;
	amount: number;
	buy_order_id: number;
	sell_order_id: number;
	amount_str: string;
	price_str: string;
	timestamp: string;
	price: number;
	type: number;
	id: number;
}

export interface IBitfinexRawTradeRest {
	timestamp: number;
	tid: number;
	price: string;
	amount: string;
	exchange: string;
	type: string;
}

export interface IGeminiTradeRest {
	timestamp: number;
	timestampms: number;
	tid: number;
	price: string;
	amount: string;
	exchange: string;
	type: string;
}

export interface IGeminiTradeWs {
	type: string;
	eventid: number;
	socket_sequence: number;
	timestamp: number;
	timestampms: number;
	events: IGeminiTradeData[];
}

export interface IGeminiTradeData {
	type: string;
	tid: number;
	price: string;
	amount: string;
	makerSide: string;
}

export type IKrakenRawTrade = IKrakenRawTradeEntry & {
	last: string;
};

export interface IKrakenRawTradeEntry {
	[pair: string]: string[][];
}

export interface IGdaxRawTradeRest {
	error: string[];
	result: {
		[code: string]: Array<Array<string | number>>;
	};
}

export interface ISubProcess {
	source: string;
	instance: ChildProcess;
	lastFailTimestamp: number;
	failCount: number;
}
