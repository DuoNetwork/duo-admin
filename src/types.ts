export interface ITrade {
	source: string;
	id: string;
	price: number;
	amount: number;
	timestamp: number;
}

export interface IPrice {
	price: number;
	volume: number;
	timestamp: number;
}

export interface IPriceBar {
	source: string;
	date: string;
	hour: string;
	minute: number;
	open: number;
	high: number;
	low: number;
	close: number;
	volume: number;
	timestamp: number;
}

export interface IEvent {
	type: string;
	id: string;
	blockHash: string;
	blockNumber: number;
	transactionHash: string;
	logStatus: string;
	parameters: {[key: string]: string};
	timestamp: number;
}

export interface IAccount {
	address: string;
	privateKey: string;
}

export interface IKey {
	publicKey: string;
	privateKey: string;
}

export interface ISqlAuth {
	host: string;
	user: string;
	password: string;
}

export interface IOption {
	live: boolean;
	server: boolean;
	dynamo: boolean;
	aws: boolean;
	gcp: boolean;
	azure: boolean;
	force: boolean;
	gasPrice: number;
	gasLimit: number;
	eth: number;
	address: string;
	addr1: string;
	addr2: string;
	privateKey: string;
	price: number;
	source: string;
	pwd: string;
	event: string;
	provider: string;
	contractState: string;
	accountNum: number;
	saveAccount: boolean;
	from: string;
	to: string;
	value: number;
	index: number;
	total: number;
	minEther: number;
	alpha: number;
	amtA: number;
	amtB: number;
	numOfMinutes: number;
	numOfHours: number;
	key: string;
	endBlk: number;
}
