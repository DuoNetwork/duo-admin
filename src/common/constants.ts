export * from '../../../duo-contract-wrapper/src/constants';
export const TRADES = 'trades';
export const COMMIT = 'commit';
export const SUBSCRIBE = 'subscribe';
export const MINUTELY = 'minutely';
export const HOURLY = 'hourly';
export const NODE = 'node';
export const CLEAN_DB = 'cleanDB';
export const FETCH_PRICE = 'fetchPrice';
export const API_BITFINEX = 'bitfinex';
export const API_GEMINI = 'gemini';
export const API_KRAKEN = 'kraken';
export const API_GDAX = 'gdax';

export const AWS_DYNAMO_API_VERSION = '2012-10-08';
// db setting
export const DB_SQL_SCHEMA_PRICEFEED = 'priceFeedDB';
export const DB_SQL_TRADE = 'trades';
export const DB_SQL_HISTORY = 'historical_price';

export const DB_DUO = 'duo';
export const DB_TRADES = 'trades';
export const DB_LIVE = 'live';
export const DB_DEV = 'dev';
export const DB_PRICES = 'prices';
export const DB_AWS_TRADES_LIVE = 'duo.trades.live';
export const DB_AWS_TRADES_DEV = 'duo.trades.dev';
export const DB_AWS_HOURLY_LIVE = 'duo.prices.60.live';
export const DB_AWS_HOURLY_DEV = 'duo.prices.60.dev';
export const DB_AWS_MINUTELY_LIVE = 'duo.prices.1.live';
export const DB_AWS_MINUTELY_DEV = 'duo.prices.1.dev';
export const DB_AWS_EVENTS_LIVE = 'duo.events.live';
export const DB_AWS_EVENTS_DEV = 'duo.events.dev';
export const DB_AWS_STATUS_LIVE = 'duo.status.live';
export const DB_AWS_STATUS_DEV = 'duo.status.dev';
export const DB_AWS_UI_EVENTS_LIVE = 'duo.uiEvents.live';
export const DB_AWS_UI_EVENTS_DEV = 'duo.uiEvents.dev';
export const DB_TX_QTE = 'quote';
export const DB_TX_BASE = 'base';
export const DB_TX_QUOTE_BASE_ID = 'quoteBaseId';
export const DB_TX_SRC = 'source';
export const DB_TX_ID = 'id';
export const DB_TX_PRICE = 'price';
export const DB_TX_AMOUNT = 'amount';
export const DB_TX_TS = 'timestamp';
export const DB_TX_SYSTIME = 'systime';
export const DB_TX_PAIR = 'pair';
export const DB_UPDATED_AT = 'updatedAt';
export const DB_TX_SRC_DHM = 'sourceDateHourMinute';
export const DB_HISTORY_PRICE = 'price';
export const DB_HISTORY_TIMESTAMP = 'timestamp';
export const DB_QUOTE_BASE_TS = 'quoteBaseTimestamp';
export const DB_HISTORY_VOLUME = 'volume';
export const DB_SRC_DHM = 'sourceDateHourMinute';
export const DB_SRC_DH = 'sourceDateHour';
export const DB_SRC_DATE = 'sourceDate';
export const DB_SRC_YM = 'sourceYearMonth';
export const DB_MN_MINUTE = 'minute';
export const DB_OHLC_OPEN = 'open';
export const DB_OHLC_HIGH = 'high';
export const DB_OHLC_LOW = 'low';
export const DB_OHLC_CLOSE = 'close';
export const DB_OHLC_VOLUME = 'volume';
export const DB_OHLC_TS = 'timestamp';
export const DB_ST_PROCESS = 'process';
export const DB_ST_TS = 'timestamp';
export const DB_ST_BLOCK = 'block';
export const DB_ST_SYSTIME = 'systime';
export const DB_EV_KEY = 'eventKey';
export const DB_EV_TIMESTAMP_ID = 'timestampId';
export const DB_EV_SYSTIME = 'systime';
export const DB_EV_BLOCK_HASH = 'blockHash';
export const DB_EV_BLOCK_NO = 'blockNumber';
export const DB_EV_TX_HASH = 'transactionHash';
export const DB_EV_LOG_STATUS = 'logStatus';
export const DB_EV_PX = 'priceInWei';
export const DB_EV_TS = 'timeInSecond';
export const DB_EV_NAV_A = 'navAInWei';
export const DB_EV_NAV_B = 'navBInWei';
export const DB_EV_TOKEN_A = 'tokenAInWei';
export const DB_EV_TOKEN_B = 'tokenBInWei';
export const DB_EV_ETH = 'ethAmtInWei';
export const DB_EV_ETH_FEE = 'ethFeeInWei';
export const DB_EV_DUO_FEE = 'duoFeeInWei';
export const DB_EV_TOTAL_SUPPLY_A = 'totalSupplyA';
export const DB_EV_TOTAL_SUPPLY_B = 'totalSupplyB';
export const DB_EV_UI_ETH = 'eth';
export const DB_EV_UI_TOKEN_A = 'tokenA';
export const DB_EV_UI_TOKEN_B = 'tokenB';
export const DB_EV_UI_ETH_FEE = 'ethFee';
export const DB_EV_UI_DUO_FEE = 'duoFee';
export const DB_STATUS_EVENT_PUBLIC_OTHERS = 'EVENT_AWS_PUBLIC_OTHERS';

export const STATUS_INTERVAL = 30;
export const TRADES_STATUS_LAST_UPDATE_INTERVAL_WS = 10;

export const AWS_DYNAMO_ROLE_TRADE = 'trade';
export const AWS_DYNAMO_ROLE_HOURLY = 'hourly';
export const AWS_DYNAMO_ROLE_MINUTELY = 'minutely';
export const AWS_DYNAMO_ROLE_EVENT = 'event';
export const AWS_DYNAMO_ROLE_STATUS = 'status';

export const API_FETCH_TRADE_INTERVAL = 1;
export const LOG_INFO = 'INFO';
export const LOG_ERROR = 'ERROR';
export const LOG_DEBUG = 'DEBUG';
export const LOG_RANKING: { [level: string]: number } = {
	[LOG_ERROR]: 0,
	[LOG_INFO]: 1,
	[LOG_DEBUG]: 2
};

export const DB_PRICES_PRIMARY_KEY_RESOLUTION: {
	[period: number]: 'minute' | 'hour' | 'day' | 'month';
} = {
	0: 'minute',
	1: 'hour',
	10: 'hour',
	60: 'day',
	360: 'month',
	1440: 'month'
};

export const API_GMN_BASE_URL = 'https://api.gemini.com';
export const API_GMN_WS_LINK = 'wss://api.gemini.com';
export const API_GMN_VERSION = '/v1';
export const API_GMN_ORDER_BOOK = '/book';
export const API_GMN_TRADE = '/trades';
export const API_BFX_BASE_URL = 'https://api.bitfinex.com';
export const API_BFX_VERSION = '/v1';
export const API_BFX_ORDER_BOOK = '/book/';
export const API_BFX_TRADE = '/trades/';
export const API_KRK_BASE_URL = 'https://api.kraken.com';
export const API_KRK_VERSION = '/0/public';
export const API_KRK_ORDER_BOOK = '/Depth';
export const API_KRK_TRADE = '/Trades';
export const API_GDAX_BASE_URL = 'https://api.gdax.com:443/products';
export const API_GDAX_TRADE = '/trades';

export const API_LIST = [API_KRAKEN, API_GEMINI, API_GDAX];
export const EXCHANGE_WEIGHTAGE_TH: { [index: number]: number[] } = {
	4: [0.35, 0.3, 0.25, 0.2],
	3: [0.55, 0.45, 0.35],
	2: [0.65, 0.5]
};
