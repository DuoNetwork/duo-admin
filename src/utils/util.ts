import moment from 'moment';
import * as CST from '../common/constants';
import { IOption } from '../common/types';
// import { Options } from 'aws-cli-js';

class Util {
	public logLevel: string = CST.LOG_INFO;

	private log(text: any, level?: string): void {
		if (level && CST.LOG_RANKING[this.logLevel] >= CST.LOG_RANKING[level])
			console.log(`${moment().format('HH:mm:ss.SSS')} [${level}]: ` + text);
	}

	public logInfo(text: any): void {
		this.log(text, CST.LOG_INFO);
	}

	public logDebug(text: any): void {
		this.log(text, CST.LOG_DEBUG);
	}

	public logError(text: any): void {
		this.log(text, CST.LOG_ERROR);
	}

	public composeQuery(paras: { [key: string]: any }): string {
		let query: string = '?';
		for (const key in paras) query += key + '=' + paras[key] + '&';
		query = query.slice(0, -1);
		return query;
	}

	public parseOptions(argv: string[]): IOption {
		const option = {
			forceREST: argv.includes('rest'),
			live: process.argv.includes('live'),
			dbLive: process.argv.includes('dbLive'),
			server: process.argv.includes('server'),
			dynamo: process.argv.includes('dynamo'),
			aws: process.argv.includes('aws'),
			gcp: process.argv.includes('gcp'),
			azure: process.argv.includes('azure'),
			force: process.argv.includes('force'),
			pair: '',
			contractType: 'Beethoven',
			tenor: 'Perpetual',
			assets: [''],
			sources: [''],
			exSources: [''],
			gasPrice: 5e9,
			gasLimit: 200000,
			source: '',
			event: '',
			provider: '',
			period: 1,
			base: 'USD',
			quote: 'ETH'
		};
		for (let i = 3; i < argv.length; i++) {
			const args = argv[i].split('=');
			switch (args[0]) {
				case 'exSources':
					option.exSources = args[1].split(',');
					break;
				case 'sources':
					option.sources = args[1].split(',');
					break;
				case 'pair':
					option.pair = args[1].replace('_', '|') || option.pair;
					break;
				case 'assets':
					option.assets = args[1].split(',');
					break;
				case 'gasPrice':
					option.gasPrice = Number(args[1]) || option.gasPrice;
					break;
				case 'gasLimit':
					option.gasLimit = Number(args[1]) || option.gasLimit;
					break;
				case 'source':
					option.source = args[1] || option.source;
					break;
				case 'event':
					option.event = args[1] || option.event;
					break;
				case 'provider':
					option.provider = args[1] || option.provider;
					break;
				case 'period':
					option.period = Number(args[1]);
					break;
				case 'base':
					option.base = args[1] || option.base;
					break;
				case 'quote':
					option.quote = args[1] || option.quote;
					break;
				case 'contractType':
					option.contractType = args[1] || option.contractType;
					break;
				case 'tenor':
					option.tenor = args[1] || option.tenor;
					break;
				default:
					break;
			}
		}

		return option;
	}

	public getPeriodStartTimestamp(timestamp: number, period: number = 1) {
		// const now = this.getUTCNowTimestamp();
		return Math.floor(timestamp / 60000 / period - 1) * 60000 * period;
	}

	public timestampToString(ts: number) {
		//just for debugging purpose
		return moment.utc(ts).format('YYYY-MM-DDTHH:mm:ss');
	}

	public getStatusProcess(tool: string, option: IOption) {
		let type = '';
		const platform = option.azure ? '_AZURE' : option.gcp ? '_GCP' : '_AWS';
		console.log('dynamo', option.dynamo);
		const privacy = option.dynamo ? '_PUBLIC' : '_PRIVATE';
		let source = '';

		switch (tool) {
			case CST.TRADES:
				type = 'TRADE';
				source = option.source ? '_' + option.source.toUpperCase() : '';
				break;
			case CST.TRIGGER:
			case CST.FETCH_EVENTS:
				type = 'EVENT';
				source =
					'_' +
					([CST.EVENT_START_PRE_RESET, CST.EVENT_START_RESET].includes(option.event)
						? option.event.toUpperCase()
						: 'OTHERS');
				break;
			case CST.COMMIT:
				type = 'FEED';
				break;
			case CST.CLEAN_DB:
				type = 'CLEANDB';
				break;
			case CST.DB_PRICES:
				type = 'PRICE';
				if (option.period === 1) source = '_MINUTELY';
				else if (option.period === 60) source = '_HOURLY';
				else source = '';
				break;
			case 'minutely':
				type = 'MINUTELY';
				break;
			case CST.FETCH_PRICE:
				type = 'FETCH';
				break;
			default:
				return '';
		}

		return type + platform + privacy + source;
	}

	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public isNumber(input: any): boolean {
		const num = Number(input);
		return isFinite(num) && !isNaN(num);
	}

	public isEmptyObject(obj: object | undefined | null): boolean {
		if (!obj) return true;

		for (const prop in obj) if (obj.hasOwnProperty(prop)) return false;

		return true;
	}

	public async sleep(ms: number) {
		return new Promise(resolve => {
			setTimeout(resolve, ms);
		});
	}
}

const util = new Util();
export default util;
