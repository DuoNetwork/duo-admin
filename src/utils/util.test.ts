import * as CST from '../common/constants';
import { IOption } from '../common/types';
import util from './util';

test('isNumber() return true for numbers', () => {
	expect(util.isNumber(5)).toBe(true);
	expect(util.isNumber(5.0)).toBe(true);
});

test('isNumber() return true for empty string and null', () => {
	expect(util.isNumber('')).toBe(true);
	expect(util.isNumber(null)).toBe(true);
});

test('isNumber() return true for number strings', () => {
	expect(util.isNumber('5')).toBe(true);
	expect(util.isNumber('5.0')).toBe(true);
});

test('isNumber() return false for other strings', () => {
	expect(util.isNumber('5.0s')).toBe(false);
	expect(util.isNumber('test')).toBe(false);
	expect(util.isNumber('NaN')).toBe(false);
});

test('isNumber() return false for undefined, infinity, NaN', () => {
	expect(util.isNumber(undefined)).toBe(false);
	expect(util.isNumber(Infinity)).toBe(false);
	expect(util.isNumber(NaN)).toBe(false);
});

test('{}, null, undefined is empty', () => {
	expect(util.isEmptyObject({})).toBe(true);
	expect(util.isEmptyObject(null)).toBe(true);
	expect(util.isEmptyObject(undefined)).toBe(true);
});

test('{test: true} is not empty', () => {
	expect(util.isEmptyObject({ test: true })).toBe(false);
});

test('composeQuery', () => {
	expect(
		util.composeQuery({
			key1: 'value1',
			key2: 'value2'
		})
	).toMatchSnapshot();
});

test('parseOptions default', () => {
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	const command = [
		'npm',
		'run',
		'trades',
		'source=',
		'pair=',
		'event=',
		'provider=',
		'contractType=',
		'tenor='
	];
	expect(util.parseOptions(command)).toMatchSnapshot();
});

test('parseOptions', () => {
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	const command = [
		'npm',
		'run',
		'trades',
		'source=source',
		'assets=quote,base',
		'period=10',
		'start=20180930T080000',
		'end=20180930T100000',
		'exSources=ex1,ex2',
		'sources=s1,s2',
		'pair=pair',
		'event=event',
		'provider=provider',
		'contractType=contractType',
		'tenor=tenor'
	];
	expect(util.parseOptions(command)).toMatchSnapshot();
});

test('getPeriodStartTimestamp', () => {
	// 1970-01-15 6:56:07
	// util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	// 1970-01-15 6:55:00
	expect(util.getPeriodStartTimestamp(1234567890)).toBe(1234500000);
	expect(util.getPeriodStartTimestamp(1234567890, 1)).toBe(1234500000);
	// 1970-01-15 6:40:00
	expect(util.getPeriodStartTimestamp(1234567890, 10)).toBe(1233600000);
	// 1970-01-15 5:00:00
	expect(util.getPeriodStartTimestamp(1234567890, 60)).toBe(1227600000);
	// 1970-01-15 0:00:00
	expect(util.getPeriodStartTimestamp(1234567890, 360)).toBe(1209600000);
	// 1970-01-14 0:00:00
	expect(util.getPeriodStartTimestamp(1234567890, 1440)).toBe(1123200000);
});

const toolToTest = [
	CST.TRADES,
	CST.COMMIT,
	CST.CLEAN_DB,
	CST.TRIGGER,
	CST.FETCH_EVENTS,
	CST.FETCH_PRICE,
	CST.DB_PRICES
];

const option: IOption = util.parseOptions(['npm', 'run', 'tool', 'period=1']);

for (const tool of toolToTest) {
	test(`getStatusProcess ${tool} useDynamo`, () => {
		option.dynamo = true;
		expect(util.getStatusProcess(tool, option)).toMatchSnapshot();
	});

	test(`getStatusProcess ${tool} do not useDynamo`, () => {
		option.dynamo = false;
		expect(util.getStatusProcess(tool, option)).toMatchSnapshot();
	});
}

test(`getStatusProcess PRICE`, () => {
	option.aws = true;
	option.gcp = false;
	option.azure = false;
	option.dynamo = true;
	option.period = 60;
	expect(util.getStatusProcess(CST.DB_PRICES, option)).toMatchSnapshot();
	option.period = 10;
	expect(util.getStatusProcess(CST.DB_PRICES, option)).toMatchSnapshot();
});

test(`getStatusProcess EVENT`, () => {
	option.gcp = true;
	option.aws = false;
	option.azure = false;
	option.dynamo = true;
	option.event = 'StartPreReset';
	expect(util.getStatusProcess(CST.FETCH_EVENTS, option)).toMatchSnapshot();
});

test(`getStatusProcess TRADES`, () => {
	option.azure = true;
	option.gcp = false;
	option.aws = false;
	option.dynamo = true;
	option.source = 'source';
	expect(util.getStatusProcess(CST.TRADES, option)).toMatchSnapshot();
});

test('log error', () => {
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	console.log = jest.fn();

	util.logLevel = CST.LOG_ERROR;
	util.logError('error');
	util.logInfo('info');
	util.logDebug('debug');
	expect((console.log as jest.Mock).mock.calls).toMatchSnapshot();
});

test('log info', () => {
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	console.log = jest.fn();

	util.logLevel = CST.LOG_INFO;
	util.logError('error');
	util.logInfo('info');
	util.logDebug('debug');
	expect((console.log as jest.Mock).mock.calls).toMatchSnapshot();
});

test('log debug', () => {
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	console.log = jest.fn();

	util.logLevel = CST.LOG_DEBUG;
	util.logError('error');
	util.logInfo('info');
	util.logDebug('debug');
	expect((console.log as jest.Mock).mock.calls).toMatchSnapshot();
});

test('sleep', async () => {
	global.setTimeout = jest.fn(resolve => resolve());
	await util.sleep(1);
	expect((global.setTimeout as jest.Mock).mock.calls).toMatchSnapshot();
});
