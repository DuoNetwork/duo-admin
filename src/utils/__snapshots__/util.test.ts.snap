// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getStatusProcess aws do not useDynamo 1`] = `"TRADE_AZURE_PRIVATE"`;

exports[`getStatusProcess aws do not useDynamo 2`] = `"FEED_AZURE_PRIVATE"`;

exports[`getStatusProcess aws do not useDynamo 3`] = `"CLEANDB_AZURE_PRIVATE"`;

exports[`getStatusProcess aws do not useDynamo 4`] = `""`;

exports[`getStatusProcess aws do not useDynamo 5`] = `"EVENT_AZURE_PRIVATE_OTHERS"`;

exports[`getStatusProcess aws do not useDynamo 6`] = `"EVENT_AZURE_PRIVATE_OTHERS"`;

exports[`getStatusProcess aws do not useDynamo 7`] = `"FETCH_AZURE_PRIVATE"`;

exports[`getStatusProcess aws do not useDynamo 8`] = `"PRICE_AZURE_PRIVATE_MINUTELY"`;

exports[`getStatusProcess aws useDynamo 1`] = `"TRADE_AZURE_PUBLIC"`;

exports[`getStatusProcess aws useDynamo 2`] = `"FEED_AZURE_PUBLIC"`;

exports[`getStatusProcess aws useDynamo 3`] = `"CLEANDB_AZURE_PUBLIC"`;

exports[`getStatusProcess aws useDynamo 4`] = `""`;

exports[`getStatusProcess aws useDynamo 5`] = `"EVENT_AZURE_PUBLIC_OTHERS"`;

exports[`getStatusProcess aws useDynamo 6`] = `"EVENT_AZURE_PUBLIC_OTHERS"`;

exports[`getStatusProcess aws useDynamo 7`] = `"FETCH_AZURE_PUBLIC"`;

exports[`getStatusProcess aws useDynamo 8`] = `"PRICE_AZURE_PUBLIC_MINUTELY"`;

exports[`getStatusProcess azure do not useDynamo 1`] = `"TRADE_AZURE_PRIVATE"`;

exports[`getStatusProcess azure do not useDynamo 2`] = `"FEED_AZURE_PRIVATE"`;

exports[`getStatusProcess azure do not useDynamo 3`] = `"CLEANDB_AZURE_PRIVATE"`;

exports[`getStatusProcess azure do not useDynamo 4`] = `""`;

exports[`getStatusProcess azure do not useDynamo 5`] = `"EVENT_AZURE_PRIVATE_OTHERS"`;

exports[`getStatusProcess azure do not useDynamo 6`] = `"EVENT_AZURE_PRIVATE_OTHERS"`;

exports[`getStatusProcess azure do not useDynamo 7`] = `"FETCH_AZURE_PRIVATE"`;

exports[`getStatusProcess azure do not useDynamo 8`] = `"PRICE_AZURE_PRIVATE_MINUTELY"`;

exports[`getStatusProcess azure useDynamo 1`] = `"TRADE_AZURE_PUBLIC"`;

exports[`getStatusProcess azure useDynamo 2`] = `"FEED_AZURE_PUBLIC"`;

exports[`getStatusProcess azure useDynamo 3`] = `"CLEANDB_AZURE_PUBLIC"`;

exports[`getStatusProcess azure useDynamo 4`] = `""`;

exports[`getStatusProcess azure useDynamo 5`] = `"EVENT_AZURE_PUBLIC_OTHERS"`;

exports[`getStatusProcess azure useDynamo 6`] = `"EVENT_AZURE_PUBLIC_OTHERS"`;

exports[`getStatusProcess azure useDynamo 7`] = `"FETCH_AZURE_PUBLIC"`;

exports[`getStatusProcess azure useDynamo 8`] = `"PRICE_AZURE_PUBLIC_MINUTELY"`;

exports[`getStatusProcess gcp do not useDynamo 1`] = `"TRADE_AZURE_PRIVATE"`;

exports[`getStatusProcess gcp do not useDynamo 2`] = `"FEED_AZURE_PRIVATE"`;

exports[`getStatusProcess gcp do not useDynamo 3`] = `"CLEANDB_AZURE_PRIVATE"`;

exports[`getStatusProcess gcp do not useDynamo 4`] = `""`;

exports[`getStatusProcess gcp do not useDynamo 5`] = `"EVENT_AZURE_PRIVATE_OTHERS"`;

exports[`getStatusProcess gcp do not useDynamo 6`] = `"EVENT_AZURE_PRIVATE_OTHERS"`;

exports[`getStatusProcess gcp do not useDynamo 7`] = `"FETCH_AZURE_PRIVATE"`;

exports[`getStatusProcess gcp do not useDynamo 8`] = `"PRICE_AZURE_PRIVATE_MINUTELY"`;

exports[`getStatusProcess gcp useDynamo 1`] = `"TRADE_AZURE_PUBLIC"`;

exports[`getStatusProcess gcp useDynamo 2`] = `"FEED_AZURE_PUBLIC"`;

exports[`getStatusProcess gcp useDynamo 3`] = `"CLEANDB_AZURE_PUBLIC"`;

exports[`getStatusProcess gcp useDynamo 4`] = `""`;

exports[`getStatusProcess gcp useDynamo 5`] = `"EVENT_AZURE_PUBLIC_OTHERS"`;

exports[`getStatusProcess gcp useDynamo 6`] = `"EVENT_AZURE_PUBLIC_OTHERS"`;

exports[`getStatusProcess gcp useDynamo 7`] = `"FETCH_AZURE_PUBLIC"`;

exports[`getStatusProcess gcp useDynamo 8`] = `"PRICE_AZURE_PUBLIC_MINUTELY"`;

exports[`parseOptions 1`] = `
Object {
  "assets": Array [
    "quote",
    "base",
  ],
  "aws": false,
  "azure": false,
  "base": "USD",
  "contractType": "Beethoven",
  "dbLive": false,
  "dynamo": false,
  "event": "",
  "exSources": Array [
    "",
  ],
  "force": false,
  "forceREST": false,
  "gasLimit": 200000,
  "gasPrice": 5000000000,
  "gcp": false,
  "live": false,
  "maturity": "Perpetual",
  "pair": "",
  "period": 10,
  "provider": "",
  "quote": "ETH",
  "server": false,
  "source": "source",
  "sources": Array [
    "",
  ],
  "ws": false,
}
`;
